### PR TITLE
Add HEALPIX reprojection

### PR DESCRIPTION
**Pull request by keflavich**
Tool to convert HEALPIX images to Galactic or ICRS coordinates, based tightly on @cdeil's function in gammapy.  See https://github.com/healpy/healpy/issues/129.

### CODE REVIEW

**Comment by cdeil** [comment]
I keep getting emails asking for this functionality (because there used to be this old function that never quite worked in Gammapy). I once again don't have time to work on this and am not very good with HEALPIX in the first place.

Some more references that might be useful:
* https://bitbucket.org/ezbc/planckpy
* http://hpixlib.readthedocs.org/en/latest/drawing.html

**Comment by astrofrog** [comment]
Sounds like something we could sprint on next week?

**Comment by cdeil** [comment]
> Sounds like something we could sprint on next week?

Definitely!
(I'm not sure how useful I'll be concerning the core implementation of the interpolation routine, but we can probably look what the others have done and I can help setting up tests using files from http://www.atnf.csiro.au/people/mcalabre/WCS/ or elsewhere.)

**Comment by lpsinger** [comment]
What is the exact call that is crashing? I'm going to try to debug it...

File: reproject/healpix.py [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+    >>> import healpy as hp
+    >>> from astropy.io import fits
+    >>> from reproject.healpix import healpix_to_image
+    >>> healpix_filename = 'healpix.fits'
+    >>> healpix_data = hp.read_map(healpix_filename)
+    >>> healpix_system = fits.getheader(healpix_filename,ext=1)['COORDSYS']
+    >>> healpix_isnested = fits.getheader(healpix_filename,ext=1)['ORDERING'] == 'NESTED'
+    >>> reference_image = fits.open('reference_image.fits')[0]
+    >>> reprojected_data = healpix_to_image(healpix_data, reference_image, healpix_system, nest=healpix_isnested)
+    >>> fits.writeto('new_image.fits', reprojected_data, reference_image.header)
```
Comment (lpsinger) [inline reproject/healpix.py]: Problem here: by default, `hp.read_map` reorders the pixels to `RING` order. I'll make a pull request to fix this...

**Comment by keflavich** [comment]
@lpsinger - if I use the incorrect sign for lat, I get a crash.  Otherwise, it works.  However, I haven't tested for the RA/Dec one reported above.

This is my incomplete / WIP test code (note that it is not a functioning test and it is missing data):
```
from astropy.io import fits
from reproject import healpix

header = fits.Header.fromtextfile('hfi_output.hdr')
url = 'http://irsa.ipac.caltech.edu/data/Planck/release_1/all-sky-maps/maps/HFI_SkyMap_217_2048_R1.10_nominal.fits'
url = 'http://www.ligo.org/scientists/first2years/2015/compare/12157/bayestar.fits.gz'
out = healpix.healpix_reproject_file('/Users/adam/Downloads/HFI_SkyMap_217_2048_R1.10_nominal_ZodiCorrected.fits', header)
```

**Comment by lpsinger** [comment]
Is reproject set up to have its doctests run as part of `setup.py test`? Can I just put the corrected example code in the docstring?

**Comment by keflavich** [comment]
Yes, I think reproject runs doctests... but I can't prove it...

**Comment by lpsinger** [comment]
> This is my incomplete / WIP test code (note that it is not a functioning test and it is missing data):

Can you provide me with a reference header?

**Comment by lpsinger** [comment]
> Yes, I think reproject runs doctests... but I can't prove it...

No, it doesn't... I asked on the Astropy list how to turn this on. Anyway, I've gotten healpix_to_image working as of 38ff431eee7c1c21e714676614b165e1ff217964. Still need to fix doctests and turn them on.

I am using this HEALPix test file:
http://www.ligo.org/scientists/first2years/2015/compare/12157/bayestar.fits.gz

Here's a plot of the original HEALPix image:
original image

And here is my test script:
```
import astropy.io.fits
import astropy.wcs
from reproject.healpix import healpix_to_image
import healpy as hp
reference_header = astropy.io.fits.Header()
reference_header.update({
 'COORDSYS': 'icrs',
 'CDELT1': -0.4,
 'CDELT2': 0.4,
 'CRPIX1': 500,
 'CRPIX2': 400,
 'CRVAL1': 180.0,
 'CRVAL2': 0.0,
 'CTYPE1': 'RA---MOL',
 'CTYPE2': 'DEC--MOL',
 'CUNIT1': 'deg',
 'CUNIT2': 'deg',
 'NAXIS': 2,
 'NAXIS1': 1000,
 'NAXIS2': 800})
healpix_filename = 'bayestar.fits.gz'
healpix_data = hp.read_map(healpix_filename, verbose=False)
healpix_coordsys = 'C'
reprojected_data = healpix_to_image(healpix_data, reference_header, healpix_coordsys, nest=False, interp=False)
astropy.io.fits.writeto('new_image.fits', reprojected_data, reference_header)
```

And here a screen shot of the output image in DS9:

ds9

**Comment by keflavich** [comment]
This is the one I was using:
```
WCSAXES =                    2 / Number of coordinate axes
CRPIX1  =              468.716 / Pixel coordinate of reference point
CRPIX2  =              1123.74 / Pixel coordinate of reference point
PC1_1   =    -0.00199999986216 / Coordinate transformation matrix element
PC2_2   =     0.00199999986216 / Coordinate transformation matrix element
CDELT1  =                  1.0 / [deg] Coordinate increment at reference point
CDELT2  =                  1.0 / [deg] Coordinate increment at reference point
CUNIT1  = 'deg'                / Units of coordinate increment and value
CUNIT2  = 'deg'                / Units of coordinate increment and value
CTYPE1  = 'GLON-CAR'           / galactic longitude, plate caree projection
CTYPE2  = 'GLAT-CAR'           / galactic latitude, plate caree projection
CRVAL1  =        3.65245551155 / [deg] Coordinate value at reference point
CRVAL2  =       0.605737090951 / [deg] Coordinate value at reference point
PV2_1   =                  0.0 / CAR projection parameter
LONPOLE =                  0.0 / [deg] Native longitude of celestial pole
LATPOLE =         89.394262909 / [deg] Native latitude of celestial pole
EQUINOX =               2000.0 / [yr] Equinox of equatorial coordinates
NAXIS   =                    2
NAXIS1  =                 3089
NAXIS2  =                 1628
COORDSYS= 'galactic'
```

**Comment by astrofrog** [comment]
For the doctests, I think there's a doctest_plus option in ``setup.cfg``? But before this all works properly, I think I'll have to merge in @cdeil's updates to the standard affiliated package files.

**Comment by bsipocz** [comment]
@astrofrog - it doesn't run atm, I can open a PR for it tomorrow.

**Comment by astrofrog** [comment]
@bsipocz - just to check, do you mean that you will open a PR to edit ``setup.cfg``? Or fix https://github.com/astrofrog/reproject/pull/57?

**Comment by astrofrog** [comment]
@lpsinger - thanks for working on this! Could you open a PR to @keflavich's branch so that we can include the commits here?

**Comment by DrPhilEvans** [comment]
This is a minor point, but when I last used the reproject option to make 
a full sky plot, my screen filled with error messages, because many of 
the pixels in the FITS image have no valid coordinates (since the FITS 
image is rectangular, but the aitoff/mollwiede/something projection is 
oval). I have a feeling this can't be suppressed without actually 
editing the healpy source code itself (I gave up and wrote my own C++ 
"HealPixToIMG" programme), but since the thread is currently active, 
it's worth asking the question: can we get rid of these "error" messages?



On 25/04/15 10:40, Thomas Robitaille wrote:
> @lpsinger <https://github.com/lpsinger> - thanks for working on this!
> Could you open a PR to @keflavich <https://github.com/keflavich>'s
> branch so that we can include the commits here?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/astrofrog/reproject/pull/27#issuecomment-96169226>.
>

-- 
-------------------------

Phil Evans,
Swift Development Scientist
X-ray and Observational Astronomy Group,
University of Leicester

Tel: +44 (0)116 252 5059
Mobile: +44 (0)7780 980240 (work)
Mobile: +44 (0)7974 977723 (personal)

pae9@leicester.ac.uk

http://www.star.le.ac.uk/~pae9
http://www.swift.ac.uk

Follow me as a Swift scientist on Twitter: @swift_phil
http://www.star.le.ac.uk/~pae9/twitter

**Comment by lpsinger** [comment]
@DrPhilEvans, did you try my branch? I'm not getting warning messages.

**Comment by DrPhilEvans** [comment]
Ah, no probably not. I got a bit lost navigating the different trees :S
Can you point me to the right place?

On 27/04/15 15:59, Leo Singer wrote:
> @DrPhilEvans <https://github.com/DrPhilEvans>, did you try my branch?
> I'm not getting warning messages.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/astrofrog/reproject/pull/27#issuecomment-96689199>.
>

-- 
-------------------------

Phil Evans,
Swift Development Scientist
X-ray and Observational Astronomy Group,
University of Leicester

Tel: +44 (0)116 252 5059
Mobile: +44 (0)7780 980240 (work)
Mobile: +44 (0)7974 977723 (personal)

pae9@leicester.ac.uk

http://www.star.le.ac.uk/~pae9
http://www.swift.ac.uk

Follow me as a Swift scientist on Twitter: @swift_phil
http://www.star.le.ac.uk/~pae9/twitter

**Comment by astrofrog** [comment]
@lpsinger - just to let you know, when you open a pull request to @keflavich's repo, this is the repo to open a PR to: https://github.com/keflavich/python-reprojection (the repo here changed name after @keflavich forked it)

**Comment by lpsinger** [comment]
> Ah, no probably not. I got a bit lost navigating the different trees :S
> Can you point me to the right place?

    $ git remote add lpsinger https://github.com/lpsinger/healpy.git
    $ git fetch lpsinger
    $ git checkout --track lpsinger/healpix

**Comment by astrofrog** [comment]
@keflavich - can you change line 96 in ``.travis.yml`` from:

```
# - if [[ $SETUP_CMD != egg_info ]]; then $PIP_INSTALL ...; fi
```

to

```
- if [[ $SETUP_CMD != egg_info ]]; then $PIP_INSTALL healpy; fi
```

? If you have no time on this you can also add me (temporarily) as a committer on your ``python-reprojection`` repo and I can try and wrap up this PR.

**Comment by lpsinger** [comment]
FYI: possibly related to using `healpy.rotator` for coordinate conversion... healpy/healpy#247

**Comment by astrofrog** [comment]
I'm going to go ahead and merge this now to take the burden off @keflavich but @lpsinger, I have some questions I'll raise in a separate issue.

**Pull request merged**